### PR TITLE
Add moveit_configs_utils package to simplify loading paramters

### PR DESCRIPTION
**Pull request by JafarAbdi**
### Description

Second attempt to simplify loading parameters

This's based on https://github.com/JafarAbdi/moveit_resources/commit/44a82c49d34bdf60ef5956bb52de57be8a660eb8

Replace: https://github.com/ros-planning/moveit2/pull/209

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py
Comment (AdamPettinger) [inline moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py:14]: I think the `str` args here should be filepaths within the package instead of just the name. In `MoveItConfigsBuilder`, this filename expands to `config/panda.urdf.xacro`. That might be OK for all the moveit config packages, but if somebody wanted to use this for their own (e.g. modified) config package, they would have to put files in specific places or modify the configs builder

File: moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py
Comment (AdamPettinger) [inline moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py:36]: We could probably convert `moveit_config` to a list here so we just need `parameters = [ moveit_config, <some_other_parameter>]`, right?

File: moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py
Comment (AdamPettinger) [inline moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py:55]: Above comment about doing `moveit_config` -> list wouldn't help here if we only want to send some of the parameters. (Side note: is there a downside passing all the moveit config parameters here?)

File: moveit_demo_nodes/run_move_group/launch/run_move_group_interface.launch.py
Comment (AdamPettinger) [inline moveit_demo_nodes/run_move_group/launch/run_move_group_interface.launch.py:16]: Calling like this seems a bit strange, but I like that it allows you to decide which functions to call, and pass parameters for each.

We should make it clear that you need `.moveit_configs()` last no matter what

**Review by AdamPettinger — COMMENTED** [review]
Overall I like this. I think there are pros/cons about this style of implementation but this (as written now) is already better than the current way of doing the same configs setup in every launch file

File: moveit_demo_nodes/run_move_group/launch/run_move_group_interface.launch.py
Comment (tylerjw) [reply moveit_demo_nodes/run_move_group/launch/run_move_group_interface.launch.py:16]: I don't know how to do it in python but the approach normally used in cpp is to make so the builder object has a conversion operator to the final type so it can be converted.  In python the types are dynamic so it means you probably need a function like this that'll return the moveit configs object.  Maybe there was some sort of to_dict() or something you could overload so the moveit_config object would just be converted into a dictionary when you use it like one?

**Comment by tylerjw** [comment]
I like the pattern here.  I wonder if there is a way to make it more generic so the sytax was more like:

```python
parameters = ParameterBuilder('robot_config_package')
  .yaml('moveit_cpp.yaml')                                                          # loads a yaml file into parameters
  .file_parameter('robot_description', 'xacro/panda.urdf.xacro')  # fills out full path based on package
  .parameter('planning_timeout', 0.1)                                         # sets a parameter
  .to_dict()
```

Then it could be used for any sort of parameters, not just a moveit config.  Then maybe there could be a moveit config builder that used the ParameterBuilder and was just a function or something like that.

**Comment by AdamPettinger** [comment]
> I like the pattern here. I wonder if there is a way to make it more generic so the sytax was more like:
> 
> ```python
> parameters = ParameterBuilder('robot_config_package')
>   .yaml('moveit_cpp.yaml')                                        # loads a yaml file into parameters
>   .file_parameter('robot_description', 'xacro/panda.urdf.xacro')  # fills out full path based on package
>   .parameter('planning_timeout', 0.1)                             # sets a parameter
>   .to_dict()
> ```
> 
> Then it could be used for any sort of parameters, not just a moveit config. Then maybe there could be a moveit config builder that used the ParameterBuilder and was just a function or something like that.

I like this a lot. It would let you do multiple inside one launch file (e.g. when getting parameters from multiple packages), and would also let you use it to return a specific set of all parameters for a config package (e.g. moveit_configs, or if some other user wants to have their own bringup/config package they could provide a `getPackageConfigs()` call)

File: moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py
Comment (JafarAbdi) [inline moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py]: Remove

File: moveit_configs_utils/setup.py
Comment (JafarAbdi) [inline moveit_configs_utils/setup.py]: This need update

File: moveit_configs_utils/package.xml
Comment (JafarAbdi) [inline moveit_configs_utils/package.xml:8]: Update

File: moveit_ros/moveit_servo/test/launch/unit_test_servo_calcs.test.py
Comment (JafarAbdi) [inline moveit_ros/moveit_servo/test/launch/unit_test_servo_calcs.test.py:15]: Use `parameter_namespace`

File: moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py
Comment (henningkayser) [inline moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py:173]: This function name is not really intuitive to me. Maybe use something like `to_config()` along with `to_dict()`?

File: moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py
Comment (henningkayser) [inline moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py:177]: Where does `self._parameters` come from?

File: moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py
Comment (henningkayser) [reply moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py:34]: So this could use `to_dict()` then, no? I like the explicitness, though.

**Review by henningkayser — COMMENTED** [review]
This is awesome, all launch files look so much cleaner! I'm just not sure if we should remove the default values from the builder functions.

Some thought: Let's say all moveit config packages provide python scripts that export their custom moveit configs. Couldn't we then simply build config hierarchies by taking another included moveit config as default and just modifying specific parts, say the kinematics yaml or something else? Given this PR it seems like a really small step to get this done. The only thing we need to watch out for is circular dependencies.

File: moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py
Comment (sjahr) [inline moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py:27]: Some planners (at least the PILZ planner) require Cartesian limits which look i.e. like this:
```
cartesian_limits:
  max_trans_vel: 1
  max_trans_acc: 2.25
  max_trans_dec: -5
  max_rot_vel: 1.57
```
Maybe you could add this as a default field too.

File: moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py
Comment (JafarAbdi) [reply moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py:173]: We decided to use `to_moveit_configs()`

File: moveit2_rolling.repos
Comment (JafarAbdi) [inline moveit2_rolling.repos:20]: Update to upstream

File: moveit2_galactic.repos
Comment (tylerjw) [inline moveit2_galactic.repos:6]: Because of how CI works now if you put something in moveit2.repos you don't also need it here.  Robert (using some shell scripting magic) made so the ros version-specific repos files extend the base one.  Putting it here might be the cause of some of the CI failures.

File: moveit2_galactic.repos
Comment (tylerjw) [inline moveit2_galactic.repos:18]: :point_up:

File: moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py
Comment (tylerjw) [inline moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py:28]: It would be nice if you could just write:

```python
parameters = [moveit_config.to_dict()]
```
or something like that?

File: moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py
Comment (tylerjw) [reply moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py:53]: Probably out of the scope of this PR but it would be really nice if the rviz plugin had the option of being pointed at the ros node that had the moveit config loaded in it (maybe even as something that could be changed in the gui) and it would use the `get_parameters` service that ros starts up to read the moveit config from that other node.

**Review by tylerjw — APPROVED** [review]
Overall I really like this PR.  I just left a couple of comments.

File: moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py
Comment (JafarAbdi) [reply moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py]: I think we should, all the configs files are associated with the urdf/srdf parameters from the `.setup_assistant` file

File: moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py
Comment (JafarAbdi) [reply moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py]: Yes, it can

File: moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py
Comment (JafarAbdi) [reply moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py]: You mean `config/kinematics.yaml` right .? I'll update the class to have them

File: moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py
Comment (JafarAbdi) [reply moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py:12]: You could pass it to the robot description function `.robot_description(file_path="config/panda.urdf.xacro", mappings={"param1": value, ....})`

File: moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py
Comment (JafarAbdi) [reply moveit_configs_utils/src/moveit_configs_utils/moveit_configs_builder.py]: Done

File: moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py
Comment (JafarAbdi) [reply moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py:28]: I'll update it now

**Review by henningkayser — COMMENTED** [review]
@JafarAbdi I think this really looks about to get ready to be merged. Before merging and adding the dependency to `parameter_tools`, I'd suggest switching to a more specific package name that distinguishes it from c++ packages.
Something like
* launch_parameter_builder
* parameter_tools_py

**Comment by JafarAbdi** [comment]
> @JafarAbdi I think this really looks about to get ready to be merged. Before merging and adding the dependency to `parameter_tools`, I'd suggest switching to a more specific package name that distinguishes it from c++ packages.
> Something like
> 
> * launch_parameter_builder
> * parameter_tools_py

The `parameter_tools` will have two packages in it, the current Python one and a C++ one

**Comment by henningkayser** [comment]
> The `parameter_tools` will have two packages in it, the current Python one and a C++ one

I know, I'm talking about the package name `parameter_builder`, not the repo (which also could use a reference to ROS 2, but that's another topic). It's just very easy to get lost in these kinds of helper packages if they have very generic names. I'm just thinking that we still have drafts for a C++ config builder API based on `ParameterSet` which should be completed in the future. Who will be able to distinguish these when all packages are named `parameter_set`, `parameter_builder`, `config_builder`, etc...
Names like `rclcpp_parameter_set`, `launch_parameter_builder` (just suggestions) would be much less confusing.

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py
Comment (sjahr) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:293]: Thanks for adding the cartesian limits! Similar to the joint limits, cartesian limits where within the robot_description_planning namespace i.e. in the [panda_moveit_config](https://github.com/ros-planning/moveit_resources/blob/master/panda_moveit_config/launch/planning_context.launch). I guess it would make sense to keep it that way for the sake of uniformity. What do you think?
```suggestion
        self.__moveit_configs.cartesian_limits = {
            self.__robot_description
            + "_planning": load_yaml(
                self._package_path / (file_path or "config/cartesian_limits.yaml")
            )
        }
```

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py
Comment (gavanderhoorn) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:41]: [PEP-8](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables) seems to suggest using double underscore prefixes for simple member variables is not recommended.

**Comment by davetcoleman** [comment]
> Using catchall names such as utils should be avoided as they do not scope what goes into the package or what should be outside the package.

Can we rename this package to something more specific, per this [ROS Convention](http://wiki.ros.org/ROS/Patterns/Conventions)?

Something like:
- moveit_config_building
- moveit_param_builder

File: moveit2.repos
Comment (davetcoleman) [inline moveit2.repos:18]: @JafarAbdi This package name is also a bit ambiguous per http://wiki.ros.org/ROS/Patterns/Conventions

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py
Comment (davetcoleman) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:7]: Can we add inline documentation at the top describing how this file is used, please? Just copy Tyler's first comment in this PR with his example code.

File: moveit_configs_utils/package.xml
Comment (davetcoleman) [inline moveit_configs_utils/package.xml:7]: Please add yourself as author @JafarAbdi

File: moveit_configs_utils/package.xml
Comment (davetcoleman) [inline moveit_configs_utils/package.xml:19]: Why is the file below empty?
> moveit_configs_utils/resource/moveit_configs_utils

File: moveit_configs_utils/setup.py
Comment (davetcoleman) [inline moveit_configs_utils/setup.py:7]: Can we make the initial version 0.0.1? For some reason in MoveIt 1, [we used 0.3.0](https://github.com/ros-planning/moveit2/blob/main/moveit_setup_assistant/templates/moveit_config_pkg_template/package.xml.template). Either way, 0.0.0 feels wrong to me.

File: moveit_configs_utils/setup.py
Comment (davetcoleman) [inline moveit_configs_utils/setup.py:17]:
```suggestion
    description="Python library for loading MoveIt config parameters in launch files",
```

File: moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py
Comment (davetcoleman) [reply moveit_demo_nodes/run_move_group/launch/run_move_group.launch.py:12]: Please add documentation for this soemwhere, perhaps in a new MoveIt tutorial? I think I would actually prefer just a section here: https://moveit.ros.org/documentation/contributing/code/

**Review by davetcoleman — COMMENTED** [review]
Great work!
Should this be a moveit-agnostic package?

File: moveit2.repos
Comment (tylerjw) [reply moveit2.repos:18]: Do you have a suggestion for a better name?  

The intent of the repo `parameter_tools` is infrastructure for the use of ros2 parameters (similar to rosparam_shortcuts was for ros1).  The reason we wanted to create a new repo is that rosparam_shortcuts was ported to ros2 but I didn't think the approach worked well to solve the new api and pain points of the ros2 parameter API.  Secondly, rosparam_shortcuts is one ros package in a repo and I wanted to have a repo with multiple ros packages as what Jafar created (parameter_builder) is a python package and what I'm working on (parameter_sets) is a C++ package that solve separate problems around the ros2 parameters.

I didn't want to conflict with the rosparam_shortcuts name and wanted a repo that contained tools for making ros2 paramters better.  A big reason why I don't want to put it in moveit2 itself is to avoid creating yet another circular dependency similar to the issues with the moveit_resources package.  moveit2 config packages (like moveit_resources) will need to depend on this package so it would be much nicer (for release testing) if this code is in a separate repo from moveit2.

**Comment by tylerjw** [comment]
> Can we rename this package to something more specific, per this [ROS Convention](http://wiki.ros.org/ROS/Patterns/Conventions)?
> 
> Something like:
> 
>     * moveit_config_building
> 
>     * moveit_param_builder

It might be worth separating parameter_builder from the moveit specific config portions of it eventually to make the more general pattern here apply outside of moveit.  One obvious flaw in the name `parameter_tools` and `parameter_builder` is that it does not tell the user that it was created for loading moveit configs.

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py
Comment (v4hn) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:215]: All file paths with explicit `/` in a string will fail on windows, won't they?
I guess you have to use a filepath class for these defaults too?

File: moveit_demo_nodes/run_move_group/launch/run_move_group_interface.launch.py
Comment (v4hn) [inline moveit_demo_nodes/run_move_group/launch/run_move_group_interface.launch.py:9]: See my other comment, but this one is on the user-side which is even worse usability from my perspective..
Not sure anything can be done about that.

File: moveit_demo_nodes/run_move_group/launch/run_move_group_interface.launch.py
Comment (v4hn) [inline moveit_demo_nodes/run_move_group/launch/run_move_group_interface.launch.py:13]: I know you might not need all parameters everywhere, but why did you decide to specify every single parameter for each use? I think I might prefer
```
MoveItConfigsBuilder("moveit_resources_panda")
.robot_description(file_path="config/panda.urdf.xacro")
.to_moveit_configs()
```
to load all other parameters with their defaults.

**Review by v4hn — COMMENTED** [review]
I just looked into this PR in more detail for the first time. This work is such a gem to make the ROS2 launch file mess usable!

Some spontaneous feedback inline.

File: moveit_demo_nodes/run_move_group/launch/run_move_group_interface.launch.py
Comment (tylerjw) [reply moveit_demo_nodes/run_move_group/launch/run_move_group_interface.launch.py:13]: :+1: to this suggestion.  I think one of the large barriers to the usability of moveit is many things don't have defaults that should.  Here is one case where the default behavior (what parts of the moveit config to load) could be defined somewhere and the builder could then just override various parts of it.

**Comment by vatanaksoytezer** [comment]
Is there anything holding us merging this?

**Comment by henningkayser** [comment]
> Is there anything holding us merging this, except for conflicts?

I think there are still several open requests and suggestions. But I agree that we should get something merged soon.

File: moveit_configs_utils/package.xml
Comment (JafarAbdi) [reply moveit_configs_utils/package.xml:20]: This's needed for ament

**Comment by JafarAbdi** [comment]
I would really appreciate a final review from you @jrgnicho @henningkayser @AndyZe @tylerjw

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py
Comment (AndyZe) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:93]:
```suggestion
        # A dictionary containing MoveItCpp related parameters.
```

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py
Comment (AndyZe) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:79]:
```suggestion
        # A dictionary that has the contents of the URDF file.
        self.robot_description = {}
        # A dictionary that has the contents of the SRDF file.
```

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py
Comment (AndyZe) [reply moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:100]: Ping to fix this

**Review by AndyZe — COMMENTED** [review]
I think this is a great idea. I've done something similar on a client project and it helped a lot.

Will this work with namespaces for dual arms? Will approve if the answer is yes.

**Review by stephanie-eng — COMMENTED** [review]
Small nit - some comments reference robot_name_moveit_config**s** as the default package name, which caused me a bit of confusion.

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py
Comment (stephanie-eng) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:147]: Add move group capabilities loader function

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py
Comment (stephanie-eng) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:262]: Support passing LaunchConfiguration as a mapping

**Review by AndyZe — APPROVED** [review]

**Comment by AndyZe** [comment]
I just spot-checked this on some of the tutorials. It still works well  :+1:

We still haven't tested it on a dual-arm system due to some unexpected issues, but I don't want to block a great PR forever. We can always fix up whatever issues emerge.

File: moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py
Comment (tylerjw) [inline moveit_configs_utils/moveit_configs_utils/moveit_configs_builder.py:57]: should you rename you package to launch_param_builder now that it has that repo name?

**Review by tylerjw — APPROVED** [review]

**Pull request merged**
